File: src/state/driftDetector.ts
import * as path from "node:path";
import type { ContainerInspect, EngineClient } from "src/services/docker/engineClient";
import type { ServiceDiff, ServiceSnapshot, ServiceSpec, StackDiff } from "src/types/stack";
import type { StateStore } from "./StateStore";
import { mergeEnv, readEnvFile } from "./envFile";
import { redactEnv } from "./secretRedactor";

const RUNTIME_ALLOWLIST = new Set([
  "PATH",
  "HOME",
  "HOSTNAME",
  "TERM",
  "LANG",
  "LC_ALL",
  "PWD",
  "SHLVL",
  "_",
]);

/**
 * Parse ports from a container's NetworkSettings.Ports map into the short
 * Compose syntax "HOST:CONTAINER[/proto]" (e.g. "3406:3306", "8080:80/tcp").
 * Only entries that have at least one HostPort binding are included.
 */
function parseActualPorts(
  networkPorts: Record<string, Array<{ HostIp: string; HostPort: string }> | null>,
): string[] {
  const result: string[] = [];
  for (const [containerPortProto, bindings] of Object.entries(networkPorts)) {
    if (!bindings || bindings.length === 0) continue;
    // containerPortProto is e.g. "3306/tcp" or "80/tcp"
    const slashIdx = containerPortProto.indexOf("/");
    const containerPort =
      slashIdx >= 0 ? containerPortProto.slice(0, slashIdx) : containerPortProto;
    const proto = slashIdx >= 0 ? containerPortProto.slice(slashIdx + 1) : "tcp";
    for (const b of bindings) {
      if (!b.HostPort) continue;
      // Omit /tcp suffix (most common) to match Compose YAML convention
      const portStr =
        proto === "tcp"
          ? `${b.HostPort}:${containerPort}`
          : `${b.HostPort}:${containerPort}/${proto}`;
      result.push(portStr);
    }
  }
  // Sort for stable comparison
  return result.sort();
}

/**
 * Normalize a volume binding for comparison.
 * Docker Compose prefixes named volumes with "<stackName>_" and always appends
 * ":rw" or ":ro" to the actual mount. Strip both so we compare apples-to-apples
 * against the user's declared spec (e.g. "db_data:/var/lib/mysql").
 */
function normalizeVolume(vol: string, stackName: string): string {
  const parts = vol.split(":");
  if (parts.length < 2) return vol;
  let source = parts[0] as string;
  const target = parts[1] as string;
  // Strip mode suffix (:rw / :ro) added by Docker
  const rest = parts.slice(2).filter((p) => p !== "rw" && p !== "ro");
  // Strip stack-name prefix added by Compose for named volumes (not bind mounts)
  const prefix = `${stackName}_`;
  if (source.startsWith(prefix)) source = source.slice(prefix.length);
  return rest.length > 0 ? `${source}:${target}:${rest.join(":")}` : `${source}:${target}`;
}

function envArrayToMap(env: string[]): Record<string, string> {
  const m: Record<string, string> = {};
  for (const line of env) {
    const i = line.indexOf("=");
    if (i < 0) continue;
    const k = line.slice(0, i);
    if (RUNTIME_ALLOWLIST.has(k)) continue;
    m[k] = line.slice(i + 1);
  }
  return m;
}

function desiredEnv(spec: ServiceSpec, cwd: string): Record<string, string> {
  const fromFile: Record<string, string> = {};
  for (const f of spec.env_file ?? []) {
    Object.assign(fromFile, readEnvFile(path.isAbsolute(f) ? f : path.join(cwd, f)));
  }
  return mergeEnv(fromFile, spec.environment ?? {});
}

function snapshot(
  image: string,
  cmd: string | string[] | null,
  ports: string[],
  envMap: Record<string, string>,
  volumes: string[],
  replicaCount: number,
  stackName: string,
  state?: string,
): ServiceSnapshot {
  return {
    image,
    command: cmd,
    ports,
    env: redactEnv(envMap, stackName),
    volumes,
    replicaCount,
    ...(state !== undefined ? { state } : {}),
  };
}

function diffSnapshots(
  desired: ServiceSnapshot,
  actual: ServiceSnapshot,
  declaredEnvKeys: Set<string>,
): Array<{ field: string; from: unknown; to: unknown }> {
  const changes: Array<{ field: string; from: unknown; to: unknown }> = [];

  // image + replicaCount: always compare
  for (const f of ["image", "replicaCount"] as const) {
    if (JSON.stringify(desired[f]) !== JSON.stringify(actual[f])) {
      changes.push({ field: f, from: desired[f], to: actual[f] });
    }
  }

  // command: only diff when the user explicitly set one in the spec.
  // If desired.command is null the image's own entrypoint/CMD is authoritative.
  if (desired.command !== null) {
    if (JSON.stringify(desired.command) !== JSON.stringify(actual.command)) {
      changes.push({ field: "command", from: desired.command, to: actual.command });
    }
  }

  // ports: compare sorted lists
  const dp = [...desired.ports].sort();
  const ap = [...actual.ports].sort();
  if (JSON.stringify(dp) !== JSON.stringify(ap)) {
    changes.push({ field: "ports", from: dp, to: ap });
  }

  // volumes: compare sorted, normalized lists
  const dv = [...desired.volumes].sort();
  const av = [...actual.volumes].sort();
  if (JSON.stringify(dv) !== JSON.stringify(av)) {
    changes.push({ field: "volumes", from: dv, to: av });
  }

  // env: only compare keys the user explicitly declared (visible)
  for (const k of declaredEnvKeys) {
    if (desired.env.secretKeys.includes(k) || actual.env.secretKeys.includes(k)) continue;
    const a = desired.env.visible[k];
    const b = actual.env.visible[k];
    if (a !== b) changes.push({ field: `env.${k}`, from: a, to: b });
  }
  // env: secret keys by presence + hash mismatch (values redacted) — only declared keys
  for (const k of desired.env.secretKeys) {
    const dh = desired.env.secretHashesByKey[k];
    const ah = actual.env.secretHashesByKey[k];
    if (dh !== ah) changes.push({ field: `env.${k}`, from: "***", to: "***" });
  }
  return changes;
}

export async function detectDrift(
  stackName: string,
  store: StateStore,
  engine: EngineClient,
  cwd: string = process.cwd(),
): Promise<StackDiff> {
  const def = store.read(stackName);
  if (!def) {
    return { stackName, status: "missing", serviceDiffs: [] };
  }
  const summaries = await engine.listContainers({
    all: true,
    filters: { label: [`com.docker.compose.project=${stackName}`] },
  });
  const inspects: ContainerInspect[] = await Promise.all(
    summaries.map((s) => engine.inspect(s.Id)),
  );
  const byService = new Map<string, ContainerInspect[]>();
  for (const c of inspects) {
    const svc = c.Config.Labels["com.docker.compose.service"];
    if (!svc) continue;
    let serviceContainers = byService.get(svc);
    if (serviceContainers === undefined) {
      serviceContainers = [];
      byService.set(svc, serviceContainers);
    }
    serviceContainers.push(c);
  }
  const desiredServices = new Set(Object.keys(def.services));
  const actualServices = new Set(byService.keys());

  const serviceDiffs: ServiceDiff[] = [];
  for (const svc of new Set([...desiredServices, ...actualServices])) {
    const spec = def.services[svc];
    const containers = byService.get(svc) ?? [];

    // Collect the env keys the user explicitly declared so we only diff those.
    const declaredEnvMap = spec ? desiredEnv(spec, cwd) : {};
    const declaredEnvKeys = new Set(Object.keys(declaredEnvMap));

    const desiredSnap = spec
      ? snapshot(
          spec.image,
          spec.command ?? null,
          [...(spec.ports ?? [])].sort(),
          declaredEnvMap,
          [...(spec.volumes ?? [])].sort(),
          spec.scale ?? 1,
          stackName,
        )
      : null;

    let actualSnap: ServiceSnapshot | null = null;
    const first = containers[0];
    if (first !== undefined) {
      const mergedEnvMap = envArrayToMap(first.Config.Env ?? []);
      const actualPorts = parseActualPorts(first.NetworkSettings.Ports);
      const actualVolumes = (first.HostConfig.Binds ?? [])
        .map((v) => normalizeVolume(v, stackName))
        .sort();
      actualSnap = snapshot(
        first.Config.Image,
        first.Config.Cmd,
        actualPorts,
        mergedEnvMap,
        actualVolumes,
        containers.length,
        stackName,
        first.State.Status,
      );
    }

    const changes =
      desiredSnap && actualSnap
        ? diffSnapshots(desiredSnap, actualSnap, declaredEnvKeys)
        : desiredSnap
          ? [{ field: "service", from: "desired", to: "missing" }]
          : [{ field: "service", from: "missing", to: "extra" }];

    serviceDiffs.push({ service: svc, desired: desiredSnap, actual: actualSnap, changes });
  }

  const allInSync = serviceDiffs.every((d) => d.changes.length === 0);
  const allDesiredMissing = serviceDiffs.every((d) => d.desired && !d.actual);
  const anyExtra = serviceDiffs.some((d) => !d.desired && d.actual);

  let status: StackDiff["status"];
  if (allInSync) status = "in_sync";
  else if (allDesiredMissing) status = "missing";
  else if (anyExtra) status = "extra";
  else status = "drift";

  return { stackName, status, serviceDiffs };
}

File: servers/docker-mcp-server/src/docker_mcp_server/state/drift_detector.py
"""Drift detection: compare desired stack YAML with live Docker containers.

Parity: ``src/state/driftDetector.ts:1-255``.

Uses the Phase-2 ``EngineClient`` Protocol only; Phase 3 supplies the real
implementation.
"""

import asyncio
from pathlib import Path
from typing import Any

from docker_mcp_server.services.docker.types import ContainerInspect, EngineClient
from docker_mcp_server.state.env_file import merge_env, read_env_file
from docker_mcp_server.state.secret_redactor import redact_env
from docker_mcp_server.state.state_store import StateStore
from docker_mcp_server.types.stack import (
    FieldChange,
    ServiceDiff,
    ServiceSnapshot,
    ServiceSpec,
    StackDiff,
)

RUNTIME_ALLOWLIST = {
    "PATH",
    "HOME",
    "HOSTNAME",
    "TERM",
    "LANG",
    "LC_ALL",
    "PWD",
    "SHLVL",
    "_",
}


def _parse_actual_ports(
    network_ports: dict[str, list[Any] | None],
) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for container_port_proto, bindings in network_ports.items():
        if not bindings:
            continue
        if "/" in container_port_proto:
            container_port, proto = container_port_proto.split("/", 1)
        else:
            container_port, proto = container_port_proto, "tcp"
        for binding in bindings:
            host_port = binding.host_port
            if not host_port:
                continue
            if proto == "tcp":
                mapping = f"{host_port}:{container_port}"
            else:
                mapping = f"{host_port}:{container_port}/{proto}"
            if mapping in seen:
                continue
            seen.add(mapping)
            out.append(mapping)
    return sorted(out)


def _normalize_volume(vol: str, stack_name: str) -> str:
    parts = vol.split(":")
    if len(parts) < 2:
        return vol
    source, target = parts[0], parts[1]
    rest = [p for p in parts[2:] if p not in ("rw", "ro")]
    prefix = f"{stack_name}_"
    if source.startswith(prefix):
        source = source[len(prefix) :]
    if rest:
        return f"{source}:{target}:{':'.join(rest)}"
    return f"{source}:{target}"


def _env_array_to_map(env: list[str]) -> dict[str, str]:
    out: dict[str, str] = {}
    for line in env:
        if "=" not in line:
            continue
        key, value = line.split("=", 1)
        if key in RUNTIME_ALLOWLIST:
            continue
        out[key] = value
    return out


def _desired_env(spec: ServiceSpec, cwd: str) -> dict[str, str]:
    from_env_file: dict[str, str] = {}
    for f in spec.env_file or []:
        p = f if Path(f).is_absolute() else str(Path(cwd) / f)
        from_env_file.update(read_env_file(p))
    return merge_env(from_env_file, spec.environment or {})


def _snapshot(
    *,
    image: str,
    command: str | list[str] | None,
    ports: list[str],
    env: dict[str, str],
    volumes: list[str],
    replica_count: int,
    stack_name: str,
    state: str | None = None,
) -> ServiceSnapshot:
    kwargs: dict[str, Any] = {
        "image": image,
        "command": command,
        "ports": ports,
        "env": redact_env(env, stack_name),
        "volumes": volumes,
        "replica_count": replica_count,
    }
    if state is not None:
        kwargs["state"] = state
    return ServiceSnapshot.model_validate(kwargs)


def _field_change(field: str, from_value: Any, to_value: Any) -> FieldChange:
    return FieldChange.model_validate(
        {"field": field, "from_": from_value, "to": to_value}
    )


def _stack_diff(
    stack_name: str, status: str, service_diffs: list[ServiceDiff]
) -> StackDiff:
    return StackDiff.model_validate(
        {
            "stack_name": stack_name,
            "status": status,
            "service_diffs": service_diffs,
        }
    )


def _diff_snapshots(
    desired: ServiceSnapshot,
    actual: ServiceSnapshot,
    declared_env_keys: set[str],
) -> list[FieldChange]:
    changes: list[FieldChange] = []

    for field in ("image", "replica_count"):
        if getattr(desired, field) != getattr(actual, field):
            changes.append(
                _field_change(field, getattr(desired, field), getattr(actual, field))
            )

    if desired.command is not None and desired.command != actual.command:
        changes.append(_field_change("command", desired.command, actual.command))

    desired_ports = sorted(desired.ports)
    actual_ports = sorted(actual.ports)
    if desired_ports != actual_ports:
        changes.append(_field_change("ports", desired_ports, actual_ports))

    desired_volumes = sorted(desired.volumes)
    actual_volumes = sorted(actual.volumes)
    if desired_volumes != actual_volumes:
        changes.append(_field_change("volumes", desired_volumes, actual_volumes))

    for key in declared_env_keys:
        if key in desired.env.secret_keys or key in actual.env.secret_keys:
            continue
        a = desired.env.visible.get(key)
        b = actual.env.visible.get(key)
        if a != b:
            changes.append(_field_change(f"env.{key}", a, b))

    for key in desired.env.secret_keys:
        a = desired.env.secret_hashes_by_key.get(key)
        b = actual.env.secret_hashes_by_key.get(key)
        if a != b:
            changes.append(_field_change(f"env.{key}", "***", "***"))

    return changes


async def detect_drift(
    stack_name: str,
    store: StateStore,
    engine: EngineClient,
    cwd: str,
) -> StackDiff:
    definition = store.read(stack_name)
    if definition is None:
        return _stack_diff(stack_name, "missing", [])

    summaries = await engine.list_containers(
        all=True,
        filters={"label": [f"com.docker.compose.project={stack_name}"]},
    )
    inspects = await asyncio.gather(*[engine.inspect(s.id) for s in summaries])

    by_service: dict[str, list[ContainerInspect]] = {}
    for insp in inspects:
        service = insp.config.labels.get("com.docker.compose.service")
        if not service:
            continue
        by_service.setdefault(service, []).append(insp)

    desired_services = set(definition.services.keys())
    actual_services = set(by_service.keys())

    service_diffs: list[ServiceDiff] = []
    for svc in sorted(desired_services | actual_services):
        spec = definition.services.get(svc)
        containers = by_service.get(svc, [])

        declared_env = _desired_env(spec, cwd) if spec else {}
        declared_env_keys = set(declared_env.keys())

        desired_snap: ServiceSnapshot | None = None
        actual_snap: ServiceSnapshot | None = None

        if spec is not None:
            desired_snap = _snapshot(
                image=spec.image,
                command=spec.command if spec.command is not None else None,
                ports=sorted(spec.ports or []),
                env=declared_env,
                volumes=sorted(spec.volumes or []),
                replica_count=spec.scale if spec.scale is not None else 1,
                stack_name=stack_name,
            )

        if containers:
            first = containers[0]
            actual_env = _env_array_to_map(first.config.env or [])
            actual_snap = _snapshot(
                image=first.config.image,
                command=first.config.cmd,
                ports=_parse_actual_ports(first.network_settings.ports or {}),
                env=actual_env,
                volumes=[
                    _normalize_volume(v, stack_name)
                    for v in (first.host_config.binds or [])
                ],
                replica_count=len(containers),
                stack_name=stack_name,
                state=first.state.status,
            )

        if desired_snap is not None and actual_snap is not None:
            changes = _diff_snapshots(desired_snap, actual_snap, declared_env_keys)
        elif desired_snap is not None:
            changes = [_field_change("service", "desired", "missing")]
        else:
            changes = [_field_change("service", "missing", "extra")]

        service_diffs.append(
            ServiceDiff(
                service=svc,
                desired=desired_snap,
                actual=actual_snap,
                changes=changes,
            )
        )

    all_in_sync = all(len(sd.changes) == 0 for sd in service_diffs)
    all_desired_missing = all(
        sd.desired is not None and sd.actual is None for sd in service_diffs
    )
    any_extra = any(sd.desired is None and sd.actual is not None for sd in service_diffs)

    if all_in_sync:
        status: Any = "in_sync"
    elif all_desired_missing:
        status = "missing"
    elif any_extra:
        status = "extra"
    else:
        status = "drift"

    return _stack_diff(stack_name, status, service_diffs)


__all__ = ["detect_drift"]

